Event color

ehannes opened the pull request:
Adds support for changing color of an event. An attribute "color" is added to the FullcalendarEngine::Event model. Fullcalendar already has this function, and this pull request enables it in this rails engine.

This pull request fixes #18

Commit:
Added color attribute to events

Commit:
Events default color same as fullcalendar's default

ehannes commented:
Will you merge this pull request?

aditya-kapoor commented on app/models/fullcalendar_engine/event_series.rb:
```diff
@@ -1,7 +1,7 @@
   class EventSeries < ActiveRecord::Base
     
-    attr_accessor :title, :description, :commit_button
+    attr_accessor :title, :description, :commit_button, :color
```
why adding `color` to `attr_accessor` list when you are proposing it to be a column of the table?

ehannes replied:
I'm sorry, but I'm not sure what you mean. I would like the user to be able to set the color of an event. To do so, shouldn't the color be accessible through attr_accessor?

aditya-kapoor replied:
I am saying that you don't need to put it in the accessor list. It will always be accessible.

ehannes replied:
Okey. I'm pretty new to Rails. Would you mind telling me what's the difference from the title and the description that's in the attr_accessor?

aditya-kapoor replied:
title and description are added as accessors in event_series model so that they can be used to set the values for any event model using that value. See [here](https://github.com/vinsol/fullcalendar-rails-engine/blob/master/app/models/fullcalendar_engine/event_series.rb#L20)
Since I don't see that you are setting the color of the event any where in the code, that's why I asked you to remove it.

ehannes replied:
Ah, good point. My intention is to add this functionality later on. Right now I'm just setting the color when the object in the controller. I never set the color attribute from the GUI. This should probably be added before merging this pull request, right?

aditya-kapoor replied:
Yes...:smile:
Can you please do that GUI part too and then we are good to go?

ehannes replied:
I could take a look at it, maybe after lunch. I'll be in touch...

aditya-kapoor replied:
You simply have to take care of the events create through the event_series part.
Just add another entry for color [here](https://github.com/vinsol/fullcalendar-rails-engine/blob/master/app/models/fullcalendar_engine/event_series.rb#L20).
And add a color picker as the input field in the form for event. You can learn more about color picker from [here](http://www.w3schools.com/tags/ref_colorpicker.asp).

ehannes replied:
So I simply add it like this?

```
self.events.create(
  :title => title,
  :description => description,
  :all_day => all_day,
  :starttime => new_start_time,
  :endtime => new_end_time,
  :color => color
)
```

aditya-kapoor replied:
Yes and add the color picker as input field in the form as I asked in one of my previous comments.

ehannes replied:
Do you know where to find such a color picker? How should you choose what color to pick?

aditya-kapoor replied:
Simply add `f.color_field :name` to the event form and you are done.

aditya-kapoor commented on app/controllers/fullcalendar_engine/events_controller.rb:
```diff
@@ -107,7 +108,7 @@ def load_event
 
     def event_params
-      params.require(:event).permit('title', 'description', 'starttime', 'endtime', 'all_day', 'period', 'frequency', 'commit_button')
+      params.require(:event).permit('title', 'description', 'starttime', 'endtime', 'all_day', 'period', 'frequency', 'commit_button', 'color')
```
can we please symbolize all the keys?
I know this is left out from our side but you can please make these keys as symbols.

ehannes replied:
You mean change 'title', 'description' etc... to :title, :description?

aditya-kapoor replied:
Yes..:+1:

ehannes replied:
Ok, I can do that :)

ehannes replied:
Now the keys are symbolized in EventController!

Commit:
Keys symbolized in EventsController

aditya-kapoor commented:
hi @ehannes where do we stand on this?

ehannes commented:
Hi!
Sorry that I've not been in touch. I used this plugin in a project that was ended in February. Fortunately, I will resume the work of that project in May then I will try to complete this pull request.

asadmoosa commented:
That is what I've been looking for. What about adding user_id? User_id + color will be awesome.  Can you please write a full project (Fullcalendar_engine + project) so that I can understand how values of color and user_id are added to MySQL. And how to retrieve calendar based on user_id + color.

ehannes commented:
This pull request will unfortunately not, if I manage to finish it, include connection to user_id. It will only include the possibility to set the color of specific events in the calendar, a feature already available in Fullcalendar. Include user_id is, in my opinion, a totally different topic.

asadmoosa commented:
Hannes, I totally agree with you that user_id is a different topic.  I see that the color attribute is added to the gem. When i create a new project and Gemfile has that calendar gem and add a new model for user.  How can the color of a speciic user be added to the event calendar?

Commit:
Colors can now be set from GUI.

ehannes commented:
@aditya-kapoor: Took me nearly a year or so, but now I've added the color attribute to the strong parameters list and a color picker to the GUI. Are we ready to merge now?

ehannes commented:
@kuldeepaggarwal would you mind checking this PR as well?

kuldeepaggarwal commented on app/controllers/fullcalendar_engine/events_controller.rb:
```diff
@@ -121,5 +122,13 @@ def determine_event_type
+        '#3B91AD' #default color
+      else
+        event.color 
+      end
```
1. can we extract default color to configuration, so that anyone can change the default as per their need?
2. a little code picking:

``` rb
def set_event_color(event)
  event.color.presence || FullcalendarEngine.config.default_color
end
```

ehannes replied:
Where would I include the default value? I had a look in `config/initializers/configuration.rb`, but if I understand it correctly, those values are passed to FullCalendar. The `default_color` value should not be passed to FullCalendar, right?

kuldeepaggarwal replied:
User should be able to create an initializer file, like devise.rb if you are aware about that. And he can set default values in that configuration file.

kuldeepaggarwal commented:
@ehannes Any updates?

ehannes commented:
@kuldeepaggarwal Not yet I'm afraid. I'm not sure how to implement the initializer file. Any links or tips about that?

kuldeepaggarwal commented:
@ehannes We will implement that later, but could you please extract that logic to `FullcalendarEngine::Configuration` class and set `default_color` in that file and use `FullcalendarEngine::Configuration.default_color`

scout208 commented:
We are using this gem for a project and we really want to use this feature. @ehannes do you have any idea when you might be finished? Or is there anything I can help with?

Commit:
Default event color moved to config file

ehannes commented:
Hi! Sorry for the delay.

@kuldeepaggarwal I gave it a try in 4bcc011. Was that what you meant? If not, please feel free to give further instructions. Otherwise I can push the branch to your repo so you can fix it.

@scout208 If we just get the default value for event color correct I think it's ready for merge.

kuldeepaggarwal commented on config/initializers/configuration.rb:
```diff
@@ -12,7 +12,8 @@
   'selectable'  => true,
-  'timeFormat'  => "h:mm t{ - h:mm t}"
+  'timeFormat'  => "h:mm t{ - h:mm t}",
+  'defaultColor' => '#3B91AD'
```
rename it to `default_color`.

ehannes replied:
All other variables in this config uses camelCase. Why should it use snake_case?

kuldeepaggarwal replied:
because all other variables are getting used on frontend, but if you just noticed at Line no: 18/19 we are using snake_case, as `mount_path` and `default_color` will be ruby realted variables.

I will think for a better implementation later.

ehannes replied:
Thanks for that clarification! I'll rename it then :)

kuldeepaggarwal commented on app/controllers/fullcalendar_engine/events_controller.rb:
```diff
@@ -121,5 +122,9 @@ def determine_event_type
     end
+
+    def set_event_color(event)
+      event.color.presence || FullcalendarEngine.config.defaultColor
```
rename it to `event_color`

kuldeepaggarwal commented:
@ehannes Just fix above issues and then I think, we are good to go.

Commit:
EventsController method set_event_color renamed

According to Ruby best practice, setter methods should not be prefixed
with 'set'.

Commit:
defaultColor renamed to default_color

Ruby related variables should use snake_case instead of camelCase.

ehannes commented:
I should probably add that I have not tested these changes. And I see that the test coverage of this engine is about 0%... I am using the engine in one of my own applications, but I haven't tested this specific branch.

kuldeepaggarwal commented:
@ehannes Can you please test this in your app? It will be a great help.

ehannes commented:
@kuldeepaggarwal Ok, I'll test it. Though, it would be advisable to add at least some basic tests to this engine.

kuldeepaggarwal commented:
I agree that there must be test cases for the engine. Will plan something for this. Can you please create an issue for this on the repo?
